File: Leetcode740. Delete and Earn/Solution.py
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        count = Counter(nums)
        nums = sorted(list(set(nums)))
        earn1, earn2 = 0,0

        for i in range(len(nums)):
            curEarn = nums[i]*count[nums[i]]

            if i>0 and nums[i]==nums[i-1]+1:
                temp=earn2
                earn2=max(curEarn+earn1, earn2)
                earn1=temp
            else:
                temp = earn2
                earn2 = curEarn + earn2
                earn1 = temp
        
        return earn2

# runtime - 0 ms
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if not nums:
            return 0
        
        max_val = max(nums)
        points = [0] * (max_val+1)

        for num in nums:
            points[num] += num
        
        prev2 = 0
        prev1 = 0

        for point in points:
            current = max(prev1, prev2+point)
            prev2 = prev1
            prev1 = current
        
        return prev1

# runtime - 1 ms using dynamic programming
class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        if not nums:
            return 0

        freq = [0] * (max(nums) + 1)
        for num in nums:
            freq[num] += num

        n = len(freq)
        dp = [0] * n
        dp[1] = freq[1]
        for i in range(2, n):
            dp[i] = max(freq[i] + dp[i-2], dp[i-1])

        return dp[n-1]

File: Leetcode740. Delete and Earn/Solution.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        int earn1=0, earn2=0;
		int n=nums.length;
		
		// count the frequency of each integer in nums
		Map<Integer, Integer>count = new HashMap<>();
		for(int num:nums) {
			count.put(num, count.getOrDefault(num, 0) + 1);
		}
		
		// Get unique numbers and sort them
        int[] uniqueNums = count.keySet().stream()
                                .mapToInt(Integer::intValue)
                                .toArray();
        Arrays.sort(uniqueNums);
		
		for(int i=0; i<uniqueNums.length; i++) {
			int curEarn = uniqueNums[i] * count.get(uniqueNums[i]);
			if(i>0 && uniqueNums[i]==uniqueNums[i-1]+1) {
				// check for the condition
				int temp = earn2;
				earn2 = Math.max(curEarn+earn1, earn2);
				earn1 = temp;
			} else {
				int temp = earn2;
				earn2 = curEarn + earn2;
				earn1 = temp;
			}
		}
		return earn2;
    }
}


// runtime - 0 ms
class Solution {
    static {
        for(int i = 0; i < 500; i++) deleteAndEarn(new int[] {1});
    }
    public static int deleteAndEarn(int[] nums) {
        int n = nums.length;
        int max = 0;
        for(int i : nums){
            if(i > max){
                max = i;
            }
        }
        int[] points = new int[max + 1];
        for(int num : nums){
            points[num] += num;
        }
        int[] dp = new int[max + 1];
        dp[1] = points[1];
        for(int i = 2; i < points.length; i++){
            int x = points[i] + dp[i - 2];
            int y = dp[i - 1];
            int res = Math.max(x, y);
            dp[i] = res;
        }
        return dp[max];
    }
}

// runtime - 2 ms
class Solution {
    public int deleteAndEarn(int[] nums) {
        int max = 0;
        for (int num : nums) {
            max = Math.max(num, max);
        }
        int[] sum = new int[max + 1];
        for (int num : nums) {
            sum[num] += num;
        }
        int[] dp = new int[max + 1];
        dp[0] = 0;
        dp[1] = sum[1];
        for (int i = 2; i <= max; i++) {
            dp[i] = Math.max(sum[i] + dp[i - 2], dp[i - 1]);
        }
        return dp[max];
    }
}
